File: sdk/typescript/src/phone/realtime/events.ts
/**
 * inkbox/phone/realtime/events.ts
 *
 * Typed observe events the platform emits on the call WebSocket when an
 * identity runs on platform-hosted voice. Field names match the wire JSON
 * (snake_case in, camelCase out); `event` is the discriminator. Every variant
 * keeps the full decoded `raw` payload so unknown fields stay reachable across
 * server versions. These frames ride the one existing per-call WebSocket and
 * each carries the `callId` it belongs to; only the outbound intervene frames
 * omit it (that socket is already scoped to one call — see `./intervene`).
 */

export interface TranscriptTurn {
  party: "local" | "remote";
  text: string;
}

export interface PostCallAction {
  action: string;
  details: unknown;
}

interface Base {
  raw: Record<string, unknown>;
}

export interface CallStartedEvent extends Base {
  event: "call.started";
  callId: string;
  agentIdentityId: string;
  direction: "inbound" | "outbound";
  phoneNumber: string | null;
}

export interface CallAnsweredEvent extends Base {
  event: "call.answered";
  callId: string;
}

export interface TranscriptEvent extends Base {
  event: "transcript";
  callId: string;
  party: "local" | "remote";
  text: string;
  isFinal: boolean;
  turnId: string | null;
}

export interface BargeInEvent extends Base {
  event: "barge_in";
  callId: string;
  turnId: string | null;
}

export interface ConsultRequestedEvent extends Base {
  event: "consult.requested";
  callId: string;
  consultId: string;
  query: string;
  transcriptTail: TranscriptTurn[];
}

export interface CallEndedEvent extends Base {
  event: "call.ended";
  callId: string;
  reason: string | null;
  postCallActions: PostCallAction[];
  transcript: TranscriptTurn[];
}

export interface UnknownEvent extends Base {
  event: string;
}

export type RealtimeEvent =
  | CallStartedEvent
  | CallAnsweredEvent
  | TranscriptEvent
  | BargeInEvent
  | ConsultRequestedEvent
  | CallEndedEvent
  | UnknownEvent;

function s(v: unknown): string {
  return typeof v === "string" ? v : v == null ? "" : String(v);
}

/** A wire string that may be absent — kept as `null` rather than "". */
function optS(v: unknown): string | null {
  return v == null ? null : s(v);
}

function turns(v: unknown): TranscriptTurn[] {
  if (!Array.isArray(v)) return [];
  return v.map((t) => ({
    party: (t as any)?.party === "local" ? "local" : "remote",
    text: s((t as any)?.text),
  }));
}

/** Decode one wire message into its typed observe event. */
export function parseEvent(d: Record<string, unknown>): RealtimeEvent {
  const raw = d;
  switch (d.event) {
    case "call.started":
      return {
        event: "call.started", raw, callId: s(d.call_id),
        agentIdentityId: s(d.agent_identity_id),
        direction: d.direction === "outbound" ? "outbound" : "inbound",
        phoneNumber: optS(d.phone_number),
      };
    case "call.answered":
      return { event: "call.answered", raw, callId: s(d.call_id) };
    case "transcript":
      return {
        event: "transcript", raw, callId: s(d.call_id),
        party: d.party === "local" ? "local" : "remote", text: s(d.text),
        isFinal: Boolean(d.is_final), turnId: optS(d.turn_id),
      };
    case "barge_in":
      return {
        event: "barge_in", raw, callId: s(d.call_id), turnId: optS(d.turn_id),
      };
    case "consult.requested":
      return {
        event: "consult.requested", raw, callId: s(d.call_id),
        consultId: s(d.consult_id), query: s(d.query),
        transcriptTail: turns(d.transcript_tail),
      };
    case "call.ended":
      return {
        event: "call.ended", raw, callId: s(d.call_id), reason: optS(d.reason),
        postCallActions: Array.isArray(d.post_call_actions)
          ? (d.post_call_actions as any[]).map((a) => ({
              action: s(a?.action), details: a?.details ?? null,
            }))
          : [],
        transcript: turns(d.transcript),
      };
    default:
      return { event: s(d.event), raw };
  }
}

File: sdk/rust/src/phone/realtime/events.rs
//! Typed observe events the platform emits on the call WebSocket when an
//! identity runs on platform-hosted voice.
//!
//! Field names match the wire JSON (snake_case); `event` is the serde tag.
//! These frames ride the one existing per-call WebSocket and each carries the
//! `call_id` it belongs to; only the outbound intervene frames omit it (that
//! socket is already scoped to one call — see [`super::intervene`]).

use serde::Deserialize;
use serde_json::Value;

/// One turn in a transcript tail / post-call transcript.
#[derive(Debug, Clone, Deserialize)]
pub struct TranscriptTurn {
    /// `"local"` (agent) or `"remote"` (caller) — matches the wire.
    pub party: String,
    pub text: String,
}

/// An action the agent registered during the call.
#[derive(Debug, Clone, Deserialize)]
pub struct PostCallAction {
    pub action: String,
    #[serde(default)]
    pub details: Value,
}

/// Server -> client observe events, discriminated on the `event` field.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "event")]
pub enum RealtimeEvent {
    #[serde(rename = "call.started")]
    CallStarted {
        call_id: String,
        agent_identity_id: String,
        /// `"inbound"` or `"outbound"`.
        direction: String,
        /// Absent on some inbound legs.
        #[serde(default)]
        phone_number: Option<String>,
    },
    #[serde(rename = "call.answered")]
    CallAnswered { call_id: String },
    #[serde(rename = "transcript")]
    Transcript {
        call_id: String,
        /// `"local"` (agent) or `"remote"` (caller).
        party: String,
        text: String,
        is_final: bool,
        #[serde(default)]
        turn_id: Option<String>,
    },
    #[serde(rename = "barge_in")]
    BargeIn {
        call_id: String,
        #[serde(default)]
        turn_id: Option<String>,
    },
    #[serde(rename = "consult.requested")]
    ConsultRequested {
        call_id: String,
        consult_id: String,
        query: String,
        #[serde(default)]
        transcript_tail: Vec<TranscriptTurn>,
    },
    #[serde(rename = "call.ended")]
    CallEnded {
        call_id: String,
        #[serde(default)]
        reason: Option<String>,
        #[serde(default)]
        post_call_actions: Vec<PostCallAction>,
        #[serde(default)]
        transcript: Vec<TranscriptTurn>,
    },
    /// An event whose `event` tag this SDK version does not model.
    #[serde(other)]
    Unknown,
}

/// Decode one wire message into its typed observe event.
pub fn parse_event(text: &str) -> serde_json::Result<RealtimeEvent> {
    serde_json::from_str(text)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_transcript_with_call_id() {
        let event = parse_event(
            r#"{"event":"transcript","call_id":"c1","party":"remote",
                "text":"hello","is_final":true,"turn_id":"t1"}"#,
        )
        .unwrap();
        match event {
            RealtimeEvent::Transcript {
                call_id,
                party,
                text,
                is_final,
                turn_id,
            } => {
                assert_eq!(call_id, "c1");
                assert_eq!(party, "remote");
                assert_eq!(text, "hello");
                assert!(is_final);
                assert_eq!(turn_id.as_deref(), Some("t1"));
            }
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn parses_call_started_with_optional_number() {
        let event = parse_event(
            r#"{"event":"call.started","call_id":"c1",
                "agent_identity_id":"id1","direction":"inbound"}"#,
        )
        .unwrap();
        match event {
            RealtimeEvent::CallStarted {
                agent_identity_id,
                direction,
                phone_number,
                ..
            } => {
                assert_eq!(agent_identity_id, "id1");
                assert_eq!(direction, "inbound");
                assert!(phone_number.is_none());
            }
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn parses_barge_in_fields() {
        // barge_in carries only call_id + optional turn_id on the hosted-call wire.
        let event = parse_event(r#"{"event":"barge_in","call_id":"c1","turn_id":null}"#).unwrap();
        match event {
            RealtimeEvent::BargeIn { call_id, turn_id } => {
                assert_eq!(call_id, "c1");
                assert!(turn_id.is_none());
            }
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn parses_consult_and_call_ended_nested() {
        let consult = parse_event(
            r#"{"event":"consult.requested","call_id":"c1","consult_id":"q1",
                "query":"refund?","transcript_tail":[{"party":"remote","text":"hi"}]}"#,
        )
        .unwrap();
        match consult {
            RealtimeEvent::ConsultRequested {
                call_id,
                transcript_tail,
                ..
            } => {
                assert_eq!(call_id, "c1");
                assert_eq!(transcript_tail[0].text, "hi");
            }
            other => panic!("unexpected: {other:?}"),
        }

        let ended = parse_event(
            r#"{"event":"call.ended","call_id":"c1","reason":"hangup",
                "post_call_actions":[{"action":"note","details":{"x":1}}],
                "transcript":[{"party":"local","text":"bye"}]}"#,
        )
        .unwrap();
        match ended {
            RealtimeEvent::CallEnded {
                call_id,
                reason,
                post_call_actions,
                transcript,
            } => {
                assert_eq!(call_id, "c1");
                assert_eq!(reason.as_deref(), Some("hangup"));
                assert_eq!(post_call_actions[0].action, "note");
                assert_eq!(transcript[0].text, "bye");
            }
            other => panic!("unexpected: {other:?}"),
        }
    }

    #[test]
    fn unknown_event_tag_maps_to_unknown() {
        let event = parse_event(r#"{"event":"future.thing","x":1}"#).unwrap();
        assert!(matches!(event, RealtimeEvent::Unknown));
    }
}
